alarm_trap_sender.cpp

```diff
@@ -50,35 +50,61 @@ AlarmFilter AlarmFilter::_instance;
 
 AlarmTrapSender AlarmTrapSender::_instance;
 
-bool ActiveAlarmList::update(AlarmTableDef& alarm_table_def, const std::string& issuer)
+bool ObservedAlarms::update(AlarmTableDef& alarm_table_def, const std::string& issuer)
 {
-  bool updated = false;
+  bool send_trap = false;
 
   std::map<unsigned int, AlarmListEntry>::iterator it = _index_to_entry.find(alarm_table_def.alarm_index());
-
+  
+  // Enter this if statement if the alarm we are currently raising is being
+  // raised with a non-CLEARED severity. 
   if (alarm_table_def.is_not_clear())
   {
+    // If either the current alarm doesn't exist in the ObservedAlarms mapping
+    // or there is an entry for the current alarm in the mapping but at a
+    // different severity to the one we are currently raising the alarm with, we
+    // want to update the mapping and send a trap for the alarm.
     if ((it == _index_to_entry.end()) || (alarm_table_def.severity() != it->second.alarm_table_def().severity()))
     {
       _index_to_entry[alarm_table_def.alarm_index()] = AlarmListEntry(alarm_table_def, issuer);
-      updated = true;
+      send_trap = true;
     }
+    // If the alarm we are currently trying to raise has an associated entry
+    // within the ObservedAlarms mapping at the same severity then we do not
+    // wish to send a trap for this alarm (as it would be a repeated message).
   }
   else
   {
-    if (it != _index_to_entry.end())
+    // If we are currently raising an alarm with CLEARED severity and it doesn't
+    // exist in the ObservedAlarms mapping we should send the trap for this
+    // alarm and add an entry to the mapping.
+    if (it == _index_to_entry.end())
     {
-      _index_to_entry.erase(it);
-      updated = true;
+      _index_to_entry[alarm_table_def.alarm_index()] = AlarmListEntry(alarm_table_def, issuer);;
+      send_trap = true;
+    }
+    else
+    {
+      // If the alarm does exist in the ObservedAlarms mapping but with a
+      // non-CLEARED severity then we want to send a trap clearing the alarm and
+      // update the mapping with the CLEARED severity.
+      if (it->second.alarm_table_def().severity() != AlarmDef::CLEARED)
+      {
+        _index_to_entry[alarm_table_def.alarm_index()] = AlarmListEntry(alarm_table_def, issuer);;
+        send_trap = true;
+      }
+      // If the alarm does exist in the ObservedAlarms mapping but with a
+      // CLEARED severity then we do not wish to send a trap for this alarm (as
+      // it would be a repeated message).
     }
   }
 
-  return updated;
+  return send_trap;
 }
 
-void ActiveAlarmList::remove(ActiveAlarmIterator& it)
+void ObservedAlarms::overwrite(ObservedAlarmsIterator& it, AlarmTableDef& alarm_table_def)
```

Comment:
Why do you need an `overwrite` method that is different from the `update` method?